Add a simple grafana-dashboard

elakito opened the pull request:
As requested during the discussion on the slack channel, I am creating this PR to show how it looks. There is a simple instruction to try it out and a screenshot. For now, all files are placed in tmp folder.

elakito commented:
@ppatierno it has been a while and I don't know if the dashboard needs to be adjusted. If some adjustment is required, I can look at it. I think it would be nice to see what Canary is doing. So, if you could review it, that would be great.
thanks, aki

ppatierno commented on tmp/canary-monitor.yaml:
```diff
@@ -0,0 +1,16 @@
+spec:
+  selector:
+    matchLabels:
+      strimzi.io/kind: Canary
```
I am not sure we should use the `strimzi.io/kind` as label for selecting the canary service. We use that label in the Strimzi world related to existing "kind" so corresponding custom resources (i.e. Kafka, KafkaConnect and so on). Maybe just using the `app: canary` label? @scholzj what's your opinion on this?

Frawless replied:
Agree, we should use label which we have in `020-Deployment.yaml`

elakito replied:
ok.

ppatierno commented on tmp/canary-monitor.yaml:
```diff
@@ -0,0 +1,16 @@
+      strimzi.io/kind: Canary
+  namespaceSelector:
+    matchNames:
+      - my-kafka
```
if this is an example you should be clear in the description that the namespace should be `my-kafka` otherwise metrics discovery will not work.

elakito replied:
Yes. I should add a comment. I thought strimzi quick-guide was also using namespace my-kafka. But it is actually using `my-kafka-project`. So, I can change it to `my-kafka-project` to align it to the quick guide add also a comment above to adjust it for the target clusters. How do you think?

ppatierno commented on tmp/canary.yaml:
```diff
@@ -0,0 +1,39 @@
+  name: canary-service
+  labels:
+    app: canary
+    strimzi.io/kind: Canary
```
see previous comment about this label.

elakito replied:
ok.

ppatierno commented on tmp/grafana-dashboards/strimzi-canary.json:
```diff
@@ -0,0 +1,1012 @@
+      "tableColumn": "",
+      "targets": [
+        {
+          "expr": "sum(strimzi_canary_records_produced_total)",
```
should it be `_consumed_` ?

elakito replied:
yes. oh. It looks like this one and for the average recording producing/consuming latency panels, the name is swapped. I'll fix them.

ppatierno commented on tmp/grafana-dashboards/strimzi-canary.json:
```diff
@@ -0,0 +1,1012 @@
+      "tableColumn": "",
+      "targets": [
+        {
+          "expr": "sum(rate(strimzi_canary_records_consumed_total[5m])*60)",
```
why reporting per mins and not per secs?

elakito replied:
For the default interval value (30-sec) used in canary, using rec/sec lead to an ugly fraction number, so I thought using rec/min would be more appropriate.

Frawless commented on tmp/canary-monitor.yaml:
```diff
@@ -0,0 +1,16 @@
+      - my-kafka
+  podMetricsEndpoints:
+  - path: /metrics
+    port: rest-api
```
```suggestion
    port: metrics
```

elakito replied:
ok.

Frawless commented on tmp/canary.yaml:
```diff
@@ -0,0 +1,39 @@
+      containers:
+      - name: canary
+        image: strimzi-canary:0.0.1
+        ports:
+        - containerPort: 8080
+          name: rest-api
```
I think we can only add this part to `packaging/install` because deployment and service is already defined there

elakito replied:
yes. When I opened this PR some months ago, the installation folder was not there, so I created this temporary deployment descriptor in this PR. I'll update it to the current state.

Frawless commented on tmp/grafana-dashboards/strimzi-canary.json:
```diff
@@ -0,0 +1,1012 @@
+        "rgba(237, 129, 40, 0.89)",
+        "#d44a3a"
+      ],
+      "datasource": "Prometheus",
```
```suggestion
      "datasource": "${DS_PROMETHEUS}",
```

we use this format in Strimzi examples so I think we should follow the same pattern

elakito replied:
ok

Frawless commented on tmp/grafana-dashboards/strimzi-canary.json:
```diff
@@ -0,0 +1,1012 @@
+      "type": "grafana",
+      "id": "grafana",
+      "name": "Grafana",
+      "version": "7.3.7"
```
this is the minimal version needed?

elakito replied:
I took these properties from the dashboards of strimzi-kafka-operator https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/grafana-dashboards/strimzi-kafka.json#L12

ppatierno commented:
@elakito do you think you can address comments from Jakub. It would be great to finally merge this one now that we had a first canary release :-)

elakito commented:
@ppatierno Ok. I'll have a look and go through with the comments...

elakito commented:
Hi all,
I adjusted the pod monitor yaml canary-monitor.yaml to reflect the comments and it works with strimzi-canary's current package/install/*.yaml files. This adjustment also deleted the tmp/canary.yaml and updated tmp/README.txt.

To finalize this PR, could you suggest where these files currently should be placed?

tmp
├── canary-monitor.yaml
├── grafana-dashboard.png
├── grafana-dashboards
    └── strimzi-canary.json

Maybe create a top level folder "metrics" and place these files there?

ppatierno commented:
@elakito first of all thank you very much for coming back to this!
About the files, I would like to follow the same pattern we have with something similar in the strimzi-kafka-operator repo.
We should have a `packaging/examples/metrics` folder with a `grafana-dashboards` subfolder where putting the `strimzi-canary.json` and maybe a `prometheus-install` subfolder containing the `canary-monitor.yaml`.
About the `tmp/README.txt` I would create a section about Monitoring in the main README file.
Not sure we really need a screenshot of the dashboard, we don't have something like this in the operator.
@scholzj what do you think about all the above?

ppatierno commented on tmp/README.txt:
```diff
@@ -0,0 +1,21 @@
+
+1. Preparation
+Make sure that you have your strimzi-canary docker image. If not, build strimzi-canary and
+make its image available to your cluster.
```
not sure we really need this pre-requisite. User can use the official canary image, not building a custom one, or?

scholzj commented:
I don't think I have a strong preference. I guess @ppatierno suggestion looks good and mirrors what we have in operators. I can also imagine the README just being in the directory with the dashboard and PodMonitor. But not strong opinions.

scholzj commented:
Thanks for the work.

@Frawless Could you please look as well since you commented before on this PR.
